### PR TITLE
[ir] MatrixField refactor 3/n: Add MatrixFieldExpression

### PR DESCRIPTION
**Pull request by strongoier**
Related issue = #5959

This PR adds definition of `MatrixFieldExpression`, which serves as the representation of a matrix field in C++. Currently only `dynamic_index_stride` has been completely moved there. Others will get migrated in future PRs.

### CODE REVIEW

File: taichi/ir/frontend_ir.h [hunk 1]
```diff
@@ -521,6 +521,24 @@ class FieldExpression : public Expression {
   TI_DEFINE_ACCEPT_FOR_EXPRESSION
 };
 
+class MatrixFieldExpression : public Expression {
+ public:
+  std::vector<Expr> fields;
+  std::vector<int> element_shape;
```
Comment (jim19930609) [inline taichi/ir/frontend_ir.h:527]: I was thinking we can infer `element_shape` from `ret_type`, whereas `ret_type` for MatrixFieldExpression should have already been set upon construction?

**Review by jim19930609 — APPROVED** [review]
LGTM!

File: taichi/ir/frontend_ir.h (hunk 1 above)
Comment (strongoier) [reply taichi/ir/frontend_ir.h:527]: This part is a bit tricky - for some use cases (like quant), the fields may have different dtypes, so the result type cannot be simply represented as `TensorType`. I plan to scalarize such tricky cases during `lower_ast`. After `lower_ast` things will work as we discussed.

**Review by AD1024 — APPROVED** [review]
LGTM :D

**Pull request merged**